File: app/static/js/maintenance.js
// /ai_rag_story_app/app/static/js/maintenance.js

/**
 * Maintenance mode notification system
 * Checks for maintenance status and shows toast notifications to users
 */

"use strict";

class MaintenanceNotifier {
    constructor() {
        this.checkInterval = 60000; // Check every minute
        this.intervalId = null;
        this.lastMaintenanceStatus = false;
        this.currentToastId = null;
    }

    /**
     * Start monitoring maintenance status
     */
    start() {
        // Check immediately on start
        this.checkMaintenanceStatus();
        
        // Set up periodic checking
        this.intervalId = setInterval(() => {
            this.checkMaintenanceStatus();
        }, this.checkInterval);
        
        console.log('Maintenance status monitoring started');
    }

    /**
     * Stop monitoring maintenance status
     */
    stop() {
        if (this.intervalId) {
            clearInterval(this.intervalId);
            this.intervalId = null;
        }
        console.log('Maintenance status monitoring stopped');
    }

    /**
     * Check maintenance status from the API
     */
    async checkMaintenanceStatus() {
        try {
            const response = await fetch('/api/v1/maintenance/status');
            if (!response.ok) {
                console.warn('Failed to check maintenance status:', response.status);
                return;
            }

            const data = await response.json();
            this.handleMaintenanceStatus(data);
        } catch (error) {
            console.warn('Error checking maintenance status:', error);
        }
    }

    /**
     * Handle maintenance status response
     */
    handleMaintenanceStatus(data) {
        const isMaintenanceActive = data.enabled;
        const message = data.message;
        const estimatedEndTime = data.estimated_end_time;

        // If maintenance just became active
        if (isMaintenanceActive && !this.lastMaintenanceStatus) {
            this.showMaintenanceToast(message, estimatedEndTime);
        }
        
        // If maintenance just ended
        else if (!isMaintenanceActive && this.lastMaintenanceStatus) {
            this.hideMaintenanceToast();
            this.showToast('🎉 System update completed! The application is now fully operational.', 'success', 8000);
        }

        this.lastMaintenanceStatus = isMaintenanceActive;
    }

    /**
     * Show maintenance notification toast
     */
    showMaintenanceToast(message, estimatedEndTime) {
        // Format the message with time info if available
        let fullMessage = message || "The application is getting an update and will be back in about 5 minutes.";
        
        if (estimatedEndTime) {
            const endTime = new Date(estimatedEndTime);
            const now = new Date();
            const minutesRemaining = Math.max(0, Math.ceil((endTime - now) / (1000 * 60)));
            
            if (minutesRemaining > 0) {
                fullMessage += ` Estimated completion: ${minutesRemaining} minute${minutesRemaining !== 1 ? 's' : ''}.`;
            }
        }

        // Show persistent toast (very long duration)
        this.showToast(
            `🔧 ${fullMessage}`,
            'warning',
            300000 // 5 minutes duration
        );

        console.log('Maintenance mode notification shown');
    }

    /**
     * Hide maintenance toast (if needed)
     */
    hideMaintenanceToast() {
        // The toast system will handle hiding automatically
        console.log('Maintenance mode ended');
    }

    /**
     * Show a toast notification (wrapper for the global showToast function)
     */
    showToast(message, type = 'info', delay = 5000) {
        if (typeof showToast === 'function') {
            showToast(message, type, delay);
        } else {
            console.warn('showToast function not available, falling back to alert');
            alert(message);
        }
    }

    /**
     * Manual trigger for admin testing
     */
    async enableMaintenanceMode(message = null, durationMinutes = 5) {
        try {
            const params = new URLSearchParams();
            if (message) params.append('message', message);
            params.append('duration_minutes', durationMinutes);

            const response = await fetch('/api/v1/maintenance/enable', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/x-www-form-urlencoded',
                },
                body: params
            });

            if (response.ok) {
                this.showToast('Maintenance mode enabled', 'success');
                // Force immediate check
                setTimeout(() => this.checkMaintenanceStatus(), 1000);
            } else {
                this.showToast('Failed to enable maintenance mode', 'error');
            }
        } catch (error) {
            console.error('Error enabling maintenance mode:', error);
            this.showToast('Error enabling maintenance mode', 'error');
        }
    }

    /**
     * Manual disable for admin testing
     */
    async disableMaintenanceMode() {
        try {
            const response = await fetch('/api/v1/maintenance/disable', {
                method: 'POST'
            });

            if (response.ok) {
                this.showToast('Maintenance mode disabled', 'success');
                // Force immediate check
                setTimeout(() => this.checkMaintenanceStatus(), 1000);
            } else {
                this.showToast('Failed to disable maintenance mode', 'error');
            }
        } catch (error) {
            console.error('Error disabling maintenance mode:', error);
            this.showToast('Error disabling maintenance mode', 'error');
        }
    }
}

// Global instance
window.maintenanceNotifier = new MaintenanceNotifier();

// Auto-start when DOM is loaded
document.addEventListener('DOMContentLoaded', function() {
    window.maintenanceNotifier.start();
});

// Expose functions for admin console testing
window.enableMaintenance = function(message, duration) {
    window.maintenanceNotifier.enableMaintenanceMode(message, duration);
};

window.disableMaintenance = function() {
    window.maintenanceNotifier.disableMaintenanceMode();
};

File: app/core/maintenance.py
# /ai_rag_story_app/app/core/maintenance.py

import json
import logging
from datetime import datetime, timedelta
from pathlib import Path
from typing import Optional, Dict, Any

logger = logging.getLogger(__name__)

# Maintenance file path
MAINTENANCE_FILE_PATH = Path("maintenance_status.json")

class MaintenanceManager:
    """Simple maintenance mode manager using a JSON file"""
    
    @staticmethod
    def set_maintenance_mode(
        enabled: bool = True,
        message: str = "The application is getting an update and will be back in about 5 minutes.",
        duration_minutes: int = 5
    ) -> None:
        """Enable or disable maintenance mode"""
        maintenance_data = {
            "enabled": enabled,
            "message": message,
            "start_time": datetime.now().isoformat() if enabled else None,
            "estimated_end_time": (datetime.now() + timedelta(minutes=duration_minutes)).isoformat() if enabled else None,
            "duration_minutes": duration_minutes if enabled else None
        }
        
        try:
            with open(MAINTENANCE_FILE_PATH, 'w') as f:
                json.dump(maintenance_data, f, indent=2)
            logger.info(f"Maintenance mode {'enabled' if enabled else 'disabled'}")
        except Exception as e:
            logger.error(f"Failed to update maintenance status: {e}")
    
    @staticmethod
    def get_maintenance_status() -> Dict[str, Any]:
        """Get current maintenance status"""
        try:
            if not MAINTENANCE_FILE_PATH.exists():
                return {"enabled": False, "message": None}
            
            with open(MAINTENANCE_FILE_PATH, 'r') as f:
                data = json.load(f)
            
            # Auto-disable if estimated end time has passed
            if data.get("enabled") and data.get("estimated_end_time"):
                estimated_end = datetime.fromisoformat(data["estimated_end_time"])
                if datetime.now() > estimated_end:
                    logger.info("Maintenance window has expired, auto-disabling maintenance mode")
                    MaintenanceManager.set_maintenance_mode(enabled=False)
                    return {"enabled": False, "message": None}
            
            return data
        except Exception as e:
            logger.error(f"Failed to read maintenance status: {e}")
            return {"enabled": False, "message": None}
    
    @staticmethod
    def is_maintenance_active() -> bool:
        """Check if maintenance mode is currently active"""
        status = MaintenanceManager.get_maintenance_status()
        return status.get("enabled", False)
    
    @staticmethod
    def get_maintenance_message() -> Optional[str]:
        """Get the current maintenance message"""
        status = MaintenanceManager.get_maintenance_status()
        return status.get("message") if status.get("enabled") else None